readthedocs/proxito/views/serve.py

```diff
@@ -274,6 +286,172 @@ def _get_canonical_redirect_type(self, request):
 
         return None
 
+    def get_using_unresolver(self, request):
+        """
+        Resolve the current request using the new proxito implementation.
+
+        This is basically a copy of the get() method,
+        but adapted to make use of the unresolved to extract the current project, version, and file.
+        """
+        unresolved_domain = request.unresolved_domain
+        # TODO: capture this path in the URL.
+        path = request.path_info
+        try:
+            unresolved = unresolver.unresolve_path(
+                unresolved_domain=unresolved_domain,
+                path=path,
+                append_indexhtml=False,
+            )
+        except (
+            VersionNotFoundError,
+            TranslationNotFoundError,
+            InvalidExternalVersionError,
+        ) as exc:
+            # TODO: find a better way to pass this to the middleware.
+            request.path_project_slug = exc.project.slug
+            raise Http404
+        except UnresolvedPathError as exc:
```

Comment:
I'm not sure to understand _why_ the flow for the redirects is managed under `UnresolvedPathError` but I don't like too much this flow.

The name of the exception is confusing. We are raising an error, but it's a common/normal/success path to follow for all canonical and system redirects. So, I'd not use the `Error` word to handle this flow.

If you are handling these cases under a try/except clause on purpose, I'd suggest at least, to change the name of the exception to `UnresolvedPathNotMatched`, meaning that it didn't match any of the regular path patterns.

Reply:
That's an invalid path for versioned documentation, we don't serve files at `docs.rtd.io/` if the project isn't a single version, so being an error makes sense to me.

In proxito, when that happens, we want to redirect to the main version, we don't want to 404.

Reply:
> That's an invalid path for versioned documentation, we don't serve files at docs.rtd.io/ if the project isn't a single version, so being an error makes sense to me.

In that case, the name of the extension could make some reference to this. The current name is confusing to me and does not communicate exactly what's the error.

If I understand correctly the case you are describing, this name would make sense `AttemptToServeFilesFromRootOnVersionedDocs`. Of course, it's too long. However, if I'm right with my interpretation of the exception, we should find a better name that communicates this clearly.

Reply:
I agree with @humitos that it's odd to have so much logic in an exception handler. I think we likely want another way to represent this, since it's actually the majority of the code here?

> In proxito, when that happens, we want to redirect to the main version, we don't want to 404.

I'm also not 100% sure this is correct. Automatic redirects have bitten us in the past, so I'd like to make sure we have a plan and documentation for what these redirects do. 

Is the idea that `docs.rtd.io/index.html` will redirect the same as `docs.rtd.io/page/index.html`?

Reply:
> Is the idea that docs.rtd.io/index.html will redirect the same as docs.rtd.io/page/index.html?

Nope, this is `docs.rtd.io/` will redirect to `docs.rtd.io/en/latest` and `docs.rtd.io/projects/subproject/` will redirect to `docs.rtd.io/projects/subproject/en/latest/`

Reply:
Changed the name to `InvalidPathForVersionedProjectError`

Reply:
This exception has a pretty explicit name 💯 . I may not even need context to have an idea what could be happening.

Reply:
> I agree with @humitos that it's odd to have so much logic in an exception handler. I think we likely want another way to represent this, since it's actually the majority of the code here?

Should we open an issue about this to consider other refactor alternatives maybe 🤔 ?

@stsewd how do you feel about our feedback? Do you also find hard to parse this code or you already have it in mind and feel it more natural?

Reply:
Logic inside this exception is pretty small now

https://github.com/readthedocs/readthedocs.org/blob/776260a47165ca43b0dfb4833b78ae5556a22421/readthedocs/proxito/views/serve.py#L331-L354

But if you mean like in general, i.e. using multiple return types instead of exceptions. The first implementation I wrote was setting some values to `None` depending on the case, and the caller would need to check if a given value was `None` or not, I found that repetitive and not that explicit. Having a single return type makes it easier, since the unresolved path/URL is valid or not, there aren't "incomplete" results.

Reply:
I'm not super worried about this flow, and the more explicit exception makes it pretty clear to me. I don't know if it's even work another issue.